sympy/matrices/common.py

```diff
@@ -2484,6 +2484,8 @@ def pow(self, exp, method=None):
 
         if method is not None and method not in ['multiply', 'mulsimp', 'jordan', 'cayley']:
             raise TypeError('No such method')
+        if exp == S.One:
```

Comment:
It can be made consistent for explicit matrices to return unevaluated MatPow object for 0, 1 if they are not square.

Reply:
What use are these unevaluated `MatPow` objects though?

I think it's cleaner just to raise for all powers of non-square matrices.

Reply:
It's because some people have used to define power of 1 or 0 for non-square matrix, so it should be up to user to substitute those symbols.

Reply:
A difficulty with this is that it also means that it should not evaluate for arbitrary symbols or other expressions in case they are 0 or 1:
```
In [8]: MatrixSymbol('A', 3, 2) ** Symbol('n', integer=True)                                                                      
---------------------------------------------------------------------------
NonSquareMatrixError
```
That means that the test:
```
if not base.is_square and exp not in [S.Zero, S.One]
```
has to become
```
if fuzzy_not(fuzzy_or(base.is_square, exp.is_zero, (exp-1).is_zero)))
```
which is less trivial to check especially where `exp` could be an expression like `sin(x)**2 + cos(x)**2`.

We could just have
```
if fuzzy_not(base.is_square)
```
which is much cleaner.

What uses are there for powers of non-square matrices. I just googled and all I found was this
https://math.stackexchange.com/a/101604/538434
(which I agree with):
```
Exponentiation of (square) matrices is a recursive function, which we define
so that `A^0 =I` and `A^1 = AA^0 = A^0A`, `A^2 = AA^1= A^1A`, *etc.*

Exponentiation of non-square matrices is not well-defined, for the simple
reason that `A^2 =AA` (and higher products of `A` with itself) is not a valid
matrix product. So, the short answer is that because we don't define
exponentiation at all for non-square matrices, even `A^0` and `A^1` are
undefined in the non-square case.

Suppose you insisted on having a definition of `A^0` and `A^1` for
`m x n` matrices for `m != n`, so that `A^1=A` and also `A^0` was some
matrix which gave you back `A` after left- or right-multiplication by `A`. But
that's absurd, because whatever you tried to define `A^0` as, only one of
those products would be defined. So we don't really have any good, unique
choice for what `A^0` should be that would also give us a reasonable
definition of `A^1` that we would expect. Lacking any good way to define
`A^0`, we leave both `A^0` and `A^1` undefined.
```

Reply:
Even defining `A**0` and `A**1` we still don't have any way of doing any sensible power arithmetic like `A**n * A**m == A ** (n + m)` so I don't see how it can be used to get anything meaningful.

Reply:
Okay, if you think that this makes the logic simpler, then I may agree on raising errors everywhere

Reply:
I think always raising an error is the most conservative choice (math-wise, not backward compatibility wise). If at a later time someone can figure out a sane and consistent set of behaviors around non-square powers that proves to be useful, this can be revisited. But the current existing support for non-square powers seems purely accidental. If this breaks something for someone, I'll certainly be very curious about their use-case.

So, who has final authority on this?

Reply:
I agree that it is probably accidental.

There is no final authority. Sympy operates by consensus. If you make the changes here then if no one objects it can be merged.